### ISSUE TITLE
Issues when using npm @elsa-workflows/elsa-workflows-studio in nested web component (using ShadowRoot)

### ISSUE THREAD

**Issue by esterkaufman** [elsa-workflows/elsa-core#2421]
Hi! 👋 
      
Firstly, thanks for your work on this project! 🙂
I'm working on a project that contains a nested web component. and the deepest child web component is the `elsa-studio-root` 

### Issues:

- All menus closed immediately when clicking to open
Since when clicking it calls the registerClickOutside which checks if context-menu contains the event.target, and in a nested web component, the **onClick.event.target** gets the parent element (wrapper web component).
the true target is in **event.path[0]** array or using the function **event.composedPath()[0]**.

To solve this, currently, as you use the _jarrvis/stencil-click-outside_ library, I used [patch-package](https://github.com/ds300/patch-package) to patch `stencil-click-outside@1.7.1` for the project I'm working on.

Here is the diff that solved my problem:

```diff
diff --git a/node_modules/stencil-click-outside/dist/esm/index.js b/node_modules/stencil-click-outside/dist/esm/index.js
index 699983b..3a3cdfd 100644
--- a/node_modules/stencil-click-outside/dist/esm/index.js
+++ b/node_modules/stencil-click-outside/dist/esm/index.js
@@ -63,7 +63,7 @@ export function removeClickOutside(component, element, callback, opt = ClickOuts
     });
 }
 function initClickOutside(event, component, element, callback, excludedNodes) {
-    const target = event.target;
+    const target = event.target == event.path[0] ? event.target : event.composedPath()[0];
     if (!element.contains(target) && !isExcluded(target, excludedNodes)) {
         callback.call(component);
     }
```
 Also, I saw in your master branch, that you replace the use of the click-outside library with this section of code:
```
@Listen('click', {target: 'window'})
  onWindowClicked(event: Event){
    const target = event.target as HTMLElement;

    if (!this.element.contains(target))
      this.closeContextMenu();
  }
```
so, great, now you can add support to a nested web component.

- designer-tree component throws a lots of errors, and ActivityPicker not opening
You are using the **d3-selection** library, when you try to find all nodes with ".node.add" / ".node.start"... classes, to register click on these nodes, in order to open the ActivityPicker.

This section of code, for example, failed in nested web components:
` const nodeStartTransform: string = d3.select('.node.start').attr('transform');`
because in d3 library they do this:
```
function select(selector) {
  return typeof selector === "string"
      ? new Selection([[document.querySelector(selector)]], [document.documentElement])
      : new Selection([[selector]], root);
}
```
but document.querySelector(selector) cannot reach the nested web components element.

> Shadow root child nodes are not returned by DOM queries even if a child node matches the given query selector
need to query them on the host component element.

Here is the workaround that solved my problem for now (as i override the global doc, i needed to support calls from outside else-studio-root to document.querySelector):
```
<elsa-studio-root server-url={this.serverUrl} ref={
      (el) => {
        document.querySelector=(selector)=>el.querySelector(selector) || document.body.querySelector(selector);
        document.querySelectorAll=(selector)=>el.querySelector(selector) || document.body.querySelectorAll(selector);        ​        ​
     ​}
```

### Reproducing Steps

1. Just open a new stencil or other compilers for the web component project
2. Add your component with `shadow: true`
3. In render() method, add the <elsa-studio-root /> component

🤞 Good Luck!

**Comment by sfmskywalker** [comment]
Hi @esterkaufman , thank you for your detailed problem description + solutions / workarounds!

Because of similar issues I faced when trying to integrate certain components, I gave up on trying to use shadow DOM. But your explanations & tips help make me understand the root issue a little bit better.

Your trick to overwrite the `document.querySelector` and `document.querySelectorAll` functions is clever; it would also solve an issue I was facing when working on the Elsa 3 designer which uses X6 (which also uses document.querySelector under the hood) - so I might give that a try and see if I can build Shadow DOM components after all :)

If you want to provide a PR that patches `tencil-click-outside` and/or overwrites the queryselector functions so that D3 plays nice too, I would welcome it :)
